desktop/core/src/desktop/lib/fs/ozone/ofs.py

```diff
@@ -0,0 +1,158 @@
+  """
+  def __init__(self, url, fs_defaultfs, logical_name=None, security_enabled=False, ssl_cert_ca_verify=True, temp_dir="/tmp", umask=0o1022):
+    super(OzoneFS, self).__init__(
+      url,
+      fs_defaultfs,
+      logical_name=logical_name,
+      security_enabled=security_enabled,
+      ssl_cert_ca_verify=ssl_cert_ca_verify,
+      temp_dir=temp_dir,
+      umask=umask
+    )
+
+    split = lib_urlparse(fs_defaultfs)
+    self._scheme = split.scheme
+    self._netloc = split.netloc
+
+    self._filebrowser_action = PERMISSION_ACTION_OFS
+    self._has_trash_support = False
+    self._is_remote = True
+
+    LOG.debug("Initializing Ozone client: %s (security: %s, superuser: %s)" % (self._url, self._security_enabled, self._superuser))
+
+  @classmethod
+  def from_config(cls, ofs_config):
+    return cls(
+        url=ofs_config.WEBHDFS_URL.get(),
+        fs_defaultfs=ofs_config.FS_DEFAULTFS.get(),
+        logical_name=ofs_config.LOGICAL_NAME.get(),
+        security_enabled=ofs_config.SECURITY_ENABLED.get(),
+        ssl_cert_ca_verify=ofs_config.SSL_CERT_CA_VERIFY.get(),
+        temp_dir=ofs_config.TEMP_DIR.get(),
+        umask=get_umask_mode(),
+    )
+
+  def strip_normpath(self, path):
+    if path.startswith('ofs://'):
+      path = path[5:]
+    elif path.startswith('ofs:/'):
+      path = path[4:]
+    return path
+
+  def normpath(self, path):
+    return normpath(path)
+
+  def netnormpath(self, path):
+    return normpath(path)
+
+  def isroot(self, path):
+    return is_root(path)
+
+  def parent_path(self, path):
+    return parent_path(path)
+
+  def listdir_stats(self, path, glob=None):
+    """
+    listdir_stats(path, glob=None) -> [ OzoneFSStat ]
+
+    Get directory listing with stats.
+    """
+    path = self.strip_normpath(path)
+    params = self._getparams()
+    if glob is not None:
+      params['filter'] = glob
+    params['op'] = 'LISTSTATUS'
+    headers = self._getheaders()
+    json = self._root.get(path, params, headers)
+    filestatus_list = json['FileStatuses']['FileStatus']
+    return [OzoneFSStat(st, path) for st in filestatus_list]
+
+  def _stats(self, path):
+    """
+    This stats method returns None if the entry is not found.
+    """
+    path = self.strip_normpath(path)
+    params = self._getparams()
+    params['op'] = 'GETFILESTATUS'
+    headers = self._getheaders()
+    try:
+      json = self._root.get(path, params, headers)
+      return OzoneFSStat(json['FileStatus'], path)
+    except WebHdfsException as ex:
+      if ex.server_exc == 'FileNotFoundException' or ex.code == 404:
+        return None
+      raise ex
+  
+  def stats(self, path):
+    """
+    stats(path) -> OzoneFSStat
+    """
+    res = self._stats(path)
+    if res is not None:
+      return res
+    raise IOError(errno.ENOENT, _("File %s not found") % path)
+
+  def filebrowser_action(self):
+    return self._filebrowser_action
+
+  def rename(self, old, new):
+    self.copy(old, new, recursive=True)
```

Comment:
Really good catch! 
Copy operations are not atomic and we missed it. We used /copy + /delete as a workaround instead of dedicated /rename API because rename was not supported for buckets and volumes in Ozone for now.

But now this doesn't look correct and we should fallback on using /rename only for rename to be atomic at-least for other paths in V1. We can address for buckets and volumes once /rename API supports it in Ozone.

Reply:
Mainly we added this because of move operation. In the move operation we are internally using rename function and ozone does not support the renaming of vol and bucket hence we used **move = copy + delete**

Reply:
We have made the changes and are now using /rename API only.